### DISCUSSION TITLE
[TypeScript] Utility type for getting the `ResultType` from a query function

### DISCUSSION THREAD

**Discussion by italodeandra** [Automattic/mongoose#10403] [Q&A] [answered]
### Actual question

Is there a utility type or a do someone have any suggestions for me to be able to get the `ResultType` from a query function?

I got to this type already: 

```ts
ReturnType<Model<User>[Method]>
```

and this actually translates to 

```ts
Query<(User & Document<any, any>)[], User & Document<any, any>, {}>
```

How do I get only the `ResultType` which is the `(User & Document<any, any>)[]` part of it?
 
### Small introduction

Hi team,

I'm building an extension of Mongoose which will transform the models you create into API's to be consumed on the frontend (just very oversimplified explanation of what it actually is).

The problem of making something in that scale, is that typings are complicated, so that's why I need to make this utility and I'm having some trouble.

**Comment by italodeandra** [comment, marked as answer]
I created this utility type which gets the thenable result

```ts
type Await<T> = T extends {
  then(onfulfilled?: (value: infer U) => unknown): unknown
}
  ? U
  : T
```

So this type get's exactly what I want: `Await<ReturnType<Model<User>[Method]>>`.

**Reply by vkarpov15** [reply]
Yep `infer` is the way to go :+1:
